renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp

```diff
@@ -663,6 +663,25 @@ void WrappedVulkan::vkUnmapMemory(VkDevice device, VkDeviceMemory mem)
     RDCASSERT(memrecord->memMapState);
     MemMapState &state = *memrecord->memMapState;
 
+    bool shouldLock = false;
+    if(state.mapCoherent)
+    {
+      SCOPED_LOCK(m_CoherentMapsLock);
+
+      int32_t idx = m_CoherentMaps.indexOf(memrecord);
+      if(idx < 0)
+        RDCERR("vkUnmapMemory for memory handle that's not currently mapped");
+      else
+        m_CoherentMaps.erase(idx);
+      shouldLock = (idx >= 0);
+    }
+
+    if(shouldLock && !state.mapFlushed)
+    {
+      state.mrLock.Lock();
+      state.mrLock.Unlock();
```

Comment:
I'm not sure I understand, this isn't guarding any code with the lock? Aside from chance there's nothing to stop this from passing here and still ending up conflicting with the flush in vkQueueSubmit.

I think we should lock after checking for capframe (to avoid any lock inversion while holding m_CapTransitionLock) and unlock after doing all the work - below `state.refData = NULL;`

Reply:
It does look weird... I could change that.
I think it works, because if it is able to lock, then there is no pending in progress (since that is protected by a lock), and at this point it already removed the record from m_CoherentMaps.

Reply:
It might work by luck in some tests, but that makes an invalid assumption that the vkQueueSubmit is already in progress and that this then blocks on the Lock() call. If this lock/unlock passes but then thread gets scheduled off and vkQueueSubmit runs past the lock then this thread could get scheduled on again and continue running without the lock.